### ISSUE TITLE
Weird illegal cyclic reference while building on 2.12

### ISSUE THREAD

**Issue by danslapman** [milessabin/shapeless#679]
I just faced really weird `illegal cyclic reference` error when build my project on scala 2.12.1

Code is here: https://github.com/danslapman/shapeless-play-json/blob/master/src/test/scala/shapelessplayjson/strict/RecordDeserializationTests.scala

It compiles without any trouble on 2.11.8

I may be a scala regression, but it also can be an issue in record macro expansion, so I'm submiting it here

**Comment by allisonhb** [comment]
@danslapman it appears to be scalac trying to double-lock a symbol (it compiles with `-Yrecursion 1` for me).

**Comment by danslapman** [comment]
@allisonhb thank You, `-Yrecursion 1` worked. Could you please explain, why it happened? Is it a regression I may report to scalac?

**Comment by milessabin** [comment]
Yes, I would very much like to know what's going on here ... it's the first time I've seen a problem of this sort.

**Comment by milessabin** [comment]
@allisonhb BTW, I would like to attribute you properly, both for your shapeless contributions, and also for the scala/scala PR of yours that I'm going to include in the next TLS releases. No problem if you'd rather not, but a PR against CONTRIBUTORS here would be very welcome.

**Comment by allisonhb** [comment]
Thanks, @milessabin! I'll PR it in if/when I get this fixed...

I'm still grubbing around in the dirt for the problem here, but it appears that something is causing scalac to do an implicit search while typing `type Book = ...`, during which it stumbles upon your `implicit val`, looks at the type, and chokes because it was in the middle of initializing the symbol for `Book`.

This compiles fine:

```
trait Reads[A] {
  def read(s: String): A
}
trait KDecl {
  import shapeless.record._
  type K = Record.`'key -> Int`.T
}
object Foo extends KDecl {
  implicit val rk : Reads[K] = (s: String) => ???
}
```

but this fails:
```
trait Reads[A] {
  def read(s: String): A
}
object Foo {
  import shapeless.record._
  type K = Record.`'key -> Int`.T
  implicit val k : Reads[K] = (s: String) => ???
}
```

My first suspicion was actually `mkTypeCarrier` for calling `c.typecheck`, but it appears that we fail even before then, which is strange. I'll let you know what else I find and if there will be a pr to here or scala/scala...

**Comment by milessabin** [comment]
Good stuff!

**Comment by allisonhb** [comment]
Heh, thanks also for the `withPropagateCyclicReferences`...

**Comment by joroKr21** [comment]
Fixed in scalac by scala/scala#5762
